Use unified BattleNet server

martincostello opened the pull request:
Implemented based on https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/issues/812#issuecomment-1789607922.

Unless it's because of my office's DNS, I couldn't get `oauth.battlenet.com.cn` to resolve, so I left the URLs for China as they were.

I don't actually have a BattleNet account so I haven't tested this against the actual server.

Resolves #812.

martincostello committed:
Use unified BattleNet server

Resolves #812.

kevinchalet commented:
> Unless it's because of my office's DNS, I couldn't get `oauth.battlenet.com.cn` to resolve, so I left the URLs for China as they were.

Yeah, I noticed the same thing when testing the OpenIddict implementation. It just seems this domain name - that has a CNAME entry that points to AWS China - is not resolvable everywhere: https://www.whatsmydns.net/#CNAME/oauth.battlenet.com.cn

kevinchalet commented on src/AspNet.Security.OAuth.BattleNet/BattleNetAuthenticationDefaults.cs:
```diff
@@ -31,6 +31,22 @@ public static class BattleNetAuthenticationDefaults
+    /// <summary>
+    /// Default value for <see cref="OAuthOptions.AuthorizationEndpoint"/>.
+    /// </summary>
+    public static readonly string AuthorizationEndpoint = "https://oauth.battle.net/oauth/authorize";
```
Instead of adding new "constants" and deprecating the other ones, we should just use the existing regions to use `oauth.battle.net`? (just in case they'd split that again... 😄)

martincostello replied:
I'll just make this internal and have all the others return this then.

To be honest, I'm not a fan of the existing implementation - a set-only property is weird.

kevinchalet replied:
> To be honest, I'm not a fan of the existing implementation - a set-only property is weird.

Yeah. If we wanted to make it look less weird, we could probably convert it to an automatic getter+setter property and move the assignation to an `IPostConfigureOptions<BattleNetOptions>` implementation.

martincostello committed:
Use IPostConfigureOptions

- Use `IPostConfigureOptions<T>` to configure the URLs for BattleNet.
- Add new `Unified` and `Custom` regions.
- Update/add XML documentation.
- Undo obsolete attributes.

kevinchalet reviewed (approved):
Looks good 👍🏻 

(adding a new `Unified` constants class is fine, but maybe that would have been safer to just update the existing "constants" to use the new server address, in case they'd change the rules in the future and "de-unify" some of the regions?)

martincostello commented:
I figured it was better to let people just switch the region so if they de-split it again, then we don't need to change anything in the code at all, people just change their `Region` value back to one of the original ones. If we change all the field values and they de-unify, then we need to do a new release to change them back again.

I figured this was a good middle ground of back-compat and not preventing the use of the unified regions at all.

kevinchalet commented:
Yeah, it's a reasonable compromise 👍🏻
